design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+Each VM’s PVCs are preserved for future scale out
+	* **Proactive** - Proactive scale-in and update by forcing VMs to shutdown during scale-in and restart during update.
+		* **ScaleInPriority** - The priority in which VM instances are selected for proactive scale-in
+			* **Random** (Default) - Pick VMs by random for scale-in first
+			* **Oldest** - Pick oldest VMs by creation time for scale-in first
+		* **ScaleInStatePreservation** - (Optional) specifies if and how to preserve state of VMs selected for scale-in.
+			* **Disabled** - (Default) all state for VMs selected for scale-in will be deleted
+			* **Offline** - PVCs for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning time during scale out)
+			* **Online** - [NOTE we can't implement this until we have the ability to suspend VM memory state to a PVC] PVCs and memory for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning and boot time during scale out)
+Each VM’s PVCs are preserved for future scale out
+* **AutohealingStrategy** - (Optional) 
+	* **None (Default)** - VM state is preserved, even during crashloop. No action is taken to auto recover the VM.
+	* **ReprovisionOnFailure** - VM is completely reprovisioned with persistent state refresh if the VM’s VMI terminates unexpectedly with VMI.Status.Phase=failed. This is useful in instances where a VM has data corruption and continues to fail liveness checks.
+
+
+## VMPool API Examples
+
+**Automatic rolling updates and "Shutdown" as scale-in strategy to optimization of boot times during scale-out**
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachinePool                                             
+metadata:                                                            
+  name: my-vm-pool                                                   
+spec:                                                                
+  virtualMachineRef:                                                                
+    name: my-template
+    kind: VirtualMachineConfig
+  replicas: 100                                                   
+  maxUnavailable: 10                                             
+  disruptionStrategy:
+    proactive:
+      scaleInStatePreservation: Offline
+```
+
+**Manual rolling updates and Manual scale-in strategy**
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachinePool                                             
+metadata:                                                            
+  name: my-vm-pool                                                   
+spec:                                                                
+  VirtualMachineConfigRef:                                                                
```

Comment:
```suggestion
  virtualMachineRef:
```
Also is the `kind` missing here? In the previous structure it is present but here not. Or here we assume that it will be set to default?

Reply:
i made this more consistent.

The assumption here is that the VirtualMachineConfigRef will be used, and just have a `name` field for now. This gives us the ability to use other API types for assigning VM configuration like data to a VMPool in the future. but for now VMCOnfigs are the only thing supported.